[WIP] Analyst Workbench

iethree opened the pull request:
## Project Info
### References
[Analyst Workbench](https://linear.app/metabase/project/analyst-workbench-d0ea0e84742b)
[Shipping Goals - Analyst Workbench](https://linear.app/metabase/document/shipping-goals-analyst-workbench-3da57ae195ae)

### Why

The Analyst Workbench is a surface within a Metabase instance that allows users to work on foundational building blocks to use for exploring their data. It will store things like Metrics, Models, and Transforms, and be used to iterate on these important entities to use in the rest of the instance.

### What’s in

- A new surface called the Workbench exists to manage building blocks for a Metabase instance
- The Workbench is used to manage the following entity types
    - Transforms
    - Models
    - Metrics
    - Table metadata
    - Model metadata
    - Segments
    - Glossary
    - SQL Snippets
    - Dependency graph
- These entities can be created, updated, and deleted from the Workbench
- Model metadata can be edited using the same UI as editing Table metadata
- Entities are no longer visible as their own menu in the sidebar navigation
    - Models
    - Metrics
    - Glossary
- Transforms and Metadata are no longer visible in the Admin panel
- The Workbench uses panels to show different levels of granularity
    - List of entities, that is sortable and filterable
    - Bottom panel for more details about a specific entity, AKA properties

iethree committed:
fix conflict

iethree committed:
fix code gen

iethree committed:
nav updates

bpander committed:
Updates workbench metric editor to use individual metric editor components rather than entire query builder

iethree committed:
tabs for transforms

bpander committed:
Stubs out using one component for both model and metric editing in workbench

iethree committed:
fix new transform creation - sort of

iethree committed:
fix rebase

iethree committed:
improve python layout etc

bpander committed:
Adds selected state to metrics and models lists in workbench

bpander committed:
Updates styles for filter/sort UI

bpander committed:
Stubs out (+) buttons for models and metrics

bpander committed:
Changes selected metric's text to brand color to match designs/models

bpander committed:
Updates metric editor in workbench so it can create new metrics (breaks workbench model editing in the process (but that needs updating anyway))

iethree committed:
make segments nice again

iethree committed:
even nicer segments

iethree committed:
even nicer segments

sloansparger committed:
config updates

sloansparger committed:
get suggested transforms working again

sloansparger committed:
rename/clean up transform editor state

bpander committed:
Stubs out a way to allow editing native models in the workbench

bpander committed:
Fix metric editor background color

iethree committed:
improve nav panel behavior

bpander committed:
Fix dark mode filter/sort controls in workbench

bpander committed:
Minor workbench model editor style tweaks

bpander committed:
Implements creating new models in workbench (native and query builder)

bpander committed:
Adds snippets to workbench

sloansparger committed:
small tweaks

iethree committed:
collapsible nav

iethree committed:
jobs and runs improvments

johnswanson committed:
Add a `transforms` application permission

This can be set just like other application permissions, e.g.:

```
curl http://localhost:3000/api/ee/advanced-permissions/application/graph

{
  "groups": {
    "2": {
      "setting":"yes",
      "monitoring":"yes",
      "subscription":"yes",
      "transforms":"yes"
    }
  },
  "revision": 1
}
```

You can read/write transforms iff you have this application level permission.

johnswanson committed:
Check new application perms in transforms API

The `GET /run` endpoint I haven't updated yet, I need to fix the read
checks for that model too I think.

In some ways it feels a bit silly to have an application level
permission but still check permissions per-object, so maybe I'll revert this.

iethree committed:
pull in glossary

iethree committed:
dependencies

iethree committed:
better debouncery

bpander committed:
Stubs out collection tree view in workbench (models only for now)

bpander committed:
Clean up AI slop

deniskaber committed:
Models metadata editing with new table UI (#64434)

* bench scaffold

* WIP: models section for workbench (#64210)

* WIP: models section for workbench

* Add more reusable component for list section

* rejigger transforms

* little fixes

* very rough cut of functioning workbench

* make routing less bad

* python tables reference fix

* rejigger transforms

* make routing less bad

* query builder in the bench

* fix conflict

* fix code gen

* nav updates

* tabs for transforms

* Updates workbench metric editor to use individual metric editor components rather than entire query builder

* Stubs out using one component for both model and metric editing in workbench

* fix new transform creation - sort of

* fix rebase

* Start with model metadata editing

* Add model field title / descr real editing

* Add model rest field props editing, moved to metadata section

* Temp

* Add models tree view in metadata section

* Add table tree selection reset on model select

* Self review fixes

---------

Co-authored-by: Ryan Laurie <iethree@gmail.com>
Co-authored-by: Sloan Sparger <sloansparger@gmail.com>
Co-authored-by: Brad Anderson <brad@metabase.com>

bpander committed:
Refactors ItemsListSection so the different entity types can have more control over their own view settings. Also makes it easier to sync those settings to url params.

deniskaber committed:
Fix models list height

bpander committed:
Fix open/close state issues in ItemsListSettings that came over from ItemsListSection

bpander committed:
Adds "Target table"/"Alphabetical" setting for transforms list

bpander committed:
Adds view setting to segments list

bpander committed:
Fixing Runs issues:
* Adds padding (UXW-1930)
* Fixes issue where filter labels are cut off
* Fixes filters

bpander committed:
Organizes metrics by collection

bpander committed:
UXW-1850 Making segments nicer

deniskaber committed:
Add missing models editing logic

bpander committed:
Removes reference to variable added in `transforms-codegen` that is required on that branch, but removed on `feature/workbench` in (9dab2ae348c)

deniskaber committed:
UXW-1976 Add sorting to the metadata of a model

bpander committed:
Adds ability to create a new job in the workbench

bpander committed:
Fix issue where not having any jobs prevents you from being able to create more jobs

bpander committed:
Updates styles for jobs list

deniskaber committed:
UXW-1970 Refactor navigation menu

bpander committed:
Fix incorrectly ported params in jobs list

deniskaber committed:
UXW-1970 Add state preserve in localStorage

bpander committed:
Consolidates onAddNewItem with AddButton props in ItemsListSection for simplicity and eliminate TS errors

bpander committed:
Fix issue where display=alphabetical is not preserved when clicking metrics and models

bpander committed:
UXW-1984 Organizes segments by db > schema > table

deniskaber committed:
UXW-1970 Fix issue with content scrolling

deniskaber committed:
UXW-1970 Fix issue with FK field options

deniskaber committed:
UXW-1970 Fix issue with horizontal scroll

bpander committed:
Tweaks styles for models, metrics, and segments trees

bpander committed:
Ports `initiallyExpanded` and `rightSection` Tree additions from `source-of-truth` branch and uses them in segments tree view

bpander committed:
Segment polish: fix issue where going from one segment's revisions to another doesn't update the revision history that's displayed. Also prevents segment tree from displaying if it's still loading table data.

bpander committed:
Wraps name in Ellipsified in ItemsListTreeNode

deniskaber committed:
Remove empty collections in models tree

bpander committed:
Fix incorrectly ported transform/job params

bpander committed:
UXW-1971 Implements tree view for transforms

bpander committed:
UXW-1972 UXW-1973 Organizes personal collections in models and metrics tree views

bpander committed:
UXW-1996 Fix issue where no + button is present when there are no transforms

bpander committed:
Fix tree view rendering issues when transform's data lacks database or schema

bpander committed:
UXW-1985 Ensures lists are sorted correctly

bpander committed:
Fixes a few TS errors

bpander committed:
Fix TS error. Reverts changes from 3631acdd17d that appear to be a result of an incorrectly resolved merge conflict.

bpander committed:
Fix verify-i18n-files error

bpander committed:
Fix low-hanging lint errors

bpander committed:
Fix jest compilation error

bpander committed:
Allows non-admins to see /bench routes

bpander committed:
Starts fixing unit tests

bpander committed:
Brings back `scheduleCallback` from original SegmentApp.jsx implementation to fix failing unit test

bpander committed:
Fix query builder unsaved changes unit tests

bpander committed:
Fix more unit tests

bpander committed:
prettier

bpander committed:
Removes unused "isEmbeddingSetup" plumbing

bpander committed:
Fix issue where metabot causes the workbench to crash for non-EE users

bpander committed:
lint-css

bpander committed:
Makes tree/alphabetical display use localStorage instead of the location query (for simplicity and because it's potentially annoying to have to switch every time you go to the bench if you prefer the non-default)

bpander committed:
UXW-1927 Adds "Last modified" option for transforms

bpander committed:
Fixes a bunch of e2e tests

bpander committed:
Fix e2e failures in group-3

bpander committed:
Fix e2e failures in group-6

bpander committed:
Fix e2e failures in group-8

bpander committed:
Fix e2e failures in group-10

bpander committed:
UXW-1861 Stubs out a "more" menu for models

bpander committed:
Fix TS error after rebase

deniskaber committed:
UXW-1977 unified metadata tree (#64751)

kamilmielnik commented on frontend/src/metabase/metadata/pages/DataModel/DataModel.tsx:
```diff
@@ -22,28 +27,32 @@ import {
+    collectionId,
+    modelId,
+    fieldName,
+  } = parseRouteParams(params);
```
Can we please have `ModelDataModel` and `TableDataModel`?
One component supporting both cases is too difficult to reason about.

bpander committed:
UXW-1862 Adds "more" menu to metrics list

bpander committed:
UXW-1861 Fix typo in copy api call

bpander committed:
UXW-2055 Adds model and metric cache settings to workbench

alxnddr committed:
update workbench nav

alxnddr committed:
fix deps

johnswanson committed:
fix tests

johnswanson committed:
fix another test

bpander committed:
UXW-2055 Adds `shouldShowQuestionSettingsSidebar` check for model's and metric's Settings tab

alxnddr committed:
uxw-2082: update nav layout

alxnddr committed:
uxw-2082: floating menu should take full height and have 16px bottom margin

bpander committed:
Updates "new native model" button to work with updates from master

bpander committed:
UXW-2080 Table picker is not shown when opening existing sql question, model, etc.

bpander committed:
Removes `showAdminSettingsItem` check for workbench link in Settings menu

bpander committed:
UXW-2054 Removes Metabot and Collapse editor buttons in Model SQL editor

bpander committed:
UXW-2078 When opening the workbench, user is taken to the last list they viewed. If user has never been to the workbench, floating nav menu is automatically opened.

deniskaber committed:
UXW 2057 update list views for all entities (#64937)

* UXW-2057 Update models and metrics

# Conflicts:
#	frontend/src/metabase/bench/components/metrics/MetricsList.tsx
#	frontend/src/metabase/bench/components/models/ModelsList.tsx

* Add transforms, jobs, segments, snippets

* Update tree views

* Update icons

* Transforms tags

* Update metadata icon colors

bpander committed:
UXW-2079 Uses fixed width for bench side nav, defaults to 320px

deniskaber committed:
Uxw 2081 update metadata side nav (#64967)

* UXW-2081 Update metadata tree nav

* Fix CI

deniskaber committed:
UXW-2100 Add a link to Metadata from a Model

bpander committed:
Adds workaround to make sure metric and model data is up-to-date when refreshing because of invalidateTags

bpander committed:
UXW-1856 Adds workbench overview page

mazameli committed:
move keyboard shortcuts nav item down in the menu

alxnddr committed:
UXW-2082: bench nav ui polish

alxnddr committed:
UXW-2097, UXW-2096: add search and virtualization to metadata, transforms, models, metrics

deniskaber committed:
Fix transforms tree view subtitle design

mazameli committed:
rearrange nav and add network icon for dep graph (#64974)

Co-authored-by: eliotmetabase <eliot@metabase.com>

ranquild committed:
Workbench transforms tabs (#64979)

ranquild committed:
python tabs

ranquild committed:
python tabs

bpander committed:
UXW-2118 Adds Open in Workbench link in models and metrics menu

ranquild committed:
python tabs

ranquild committed:
Merge remote-tracking branch 'origin/feature/workbench' into feature/workbench

# Conflicts:

ranquild committed:
python tabs

ranquild committed:
Fix layout

ranquild committed:
Fix layout

ranquild committed:
Fix types

ranquild committed:
Fix initial state

ranquild committed:
Fix jumping

ranquild committed:
Fix metadata link

alxnddr committed:
UXW-2116: metadata search should include models

bpander committed:
Remove duplicated snippets link

alxnddr committed:
UXW-2083: update bench navbar

ranquild committed:
Runs

ranquild committed:
Runs

ranquild committed:
Merge remote-tracking branch 'origin/feature/workbench' into feature/workbench

johnswanson committed:
fix more backend tests

bpander committed:
Adds icons to BenchTabs

deniskaber committed:
UXW-2084 Restrict access to Transforms, Metadata and Python Library (#65007)

# Conflicts:
#	enterprise/frontend/src/metabase-enterprise/transforms/routes.tsx
#	frontend/src/metabase/bench/components/BenchNavMenu.tsx
#	frontend/src/metabase/bench/constants/navigation.ts

deniskaber committed:
UXW-2084 fixes for types and Python transforms

deniskaber committed:
Fix Transforms list item size

bpander committed:
UXW-2086 Disallow creating SQL entities if user doesn't have create queries permissions

bpander committed:
Updates Overview page so it uses the same permissions-based show/hide logic as the hover nav

sloansparger committed:
Merge branch 'master' into feature/workbench

alxnddr committed:
add padding

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
Add dependencies

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

deniskaber committed:
UXW-2131 main section nav hover state color

ranquild committed:
wip

ranquild committed:
Merge remote-tracking branch 'origin/feature/workbench' into feature/workbench

# Conflicts:

ranquild committed:
wip

ranquild committed:
wip

ranquild committed:
wip

bpander committed:
Starts adding editable title for models/metrics

ranquild committed:
Fix run button state

ranquild committed:
Merge remote-tracking branch 'origin/feature/workbench' into feature/workbench

ranquild committed:
Fix node layout

ranquild committed:
merge

ranquild committed:
loaders

ranquild committed:
paddings

ranquild committed:
paddings

ranquild committed:
paddings

ranquild committed:
Fix

bpander committed:
Adds info button and sidebar to model editor

bpander committed:
UXW-2120 Updates name on blur for models in workbench

bpander committed:
UXW-2120 Handles changing name input in new models

bpander committed:
Removes now-unnecessary invalidateTags calls

bpander committed:
UXW-2120 Makes model info button work more like chill mode model editor (only present when not dirty to prevent data from being overwritten)

sloansparger committed:
Fix transforms codegen feature in workbench (#65070)

* fix transform codegen feature + related e2e tests

* rip out admin metabot button

* impl automated pr feedback

deniskaber committed:
Uxw 2130 polish lists UI (#65075)

* UXW-2130 Polish lists UI

* Remove automatic menu open on first bench launch

deniskaber committed:
Fix typecheck

deniskaber committed:
Merge branch 'master' into feature/workbench

# Conflicts:
#	frontend/src/metabase/querying/metrics/components/MetricEditor/MetricEditor.tsx

bpander committed:
UXW-2120 Adds info button for metrics, makes it so they can be renamed

alxnddr committed:
tiny ui polish

ranquild committed:
Fix

ranquild committed:
Fix

ranquild committed:
Merge remote-tracking branch 'origin/feature/workbench' into feature/workbench

ranquild committed:
Fix

ranquild committed:
Fix

ranquild committed:
Fix

ranquild committed:
Fix

bpander committed:
UXW-2152 Adds tabs slot to BenchPaneHeader for better alignment, adds withBorder to models/metrics instances

bpander committed:
UXW-2152 Adds cancel buttons for models and metrics

bpander committed:
UXW-2155 Hides "Add" button for models and metrics if user does not have data access

johnswanson committed:
Fix tests
